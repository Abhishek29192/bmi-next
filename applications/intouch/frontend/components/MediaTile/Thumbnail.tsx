import { BMI, Icon } from "@bmi-digital/components";
import { Folder, PlayArrow } from "@material-ui/icons";
import React from "react";
import { MediaItem } from "../../lib/media/types";
import { getVimeoEmbedUrl, resizeContentfulImage } from "../../lib/media/utils";
import styles from "./styles.module.scss";

type Props = {
  mediaItem: MediaItem;
  optanonClass?: string;
};

const THUMBNAIL_MAX_WIDTH = 300;

const VimeoIFrameThumbnail = ({
  videoUrl,
  optanonClass,
  mediaItemClass
}: {
  videoUrl: string;
  optanonClass: string;
  mediaItemClass: string;
}) => (
  <div>
    <iframe
      src={`${getVimeoEmbedUrl(
        videoUrl
      )}?title=0&portrait=0&controls=0&sidedock=0`}
      height="100%"
      width="100%"
      scrolling="yes"
      frameBorder="0"
      className={`${styles.iframe} ${
        mediaItemClass?.length ? mediaItemClass : optanonClass
      }`}
      sandbox="allow-scripts allow-same-origin"
    />

    <div className={styles.playButton}>
      <Icon
        source={PlayArrow}
        style={{ fontSize: "2rem", display: "block" }}
        color="primary"
      />
    </div>
  </div>
);

export const MediaTileThumbnail = ({ mediaItem, optanonClass }: Props) => {
  if (mediaItem.__typename === "MediaFolder") {
    return <Icon source={Folder} style={{ fontSize: 64 }} color="primary" />;
  }
  // precedence always to thumbnails uploaded by the market admin
  if (mediaItem.thumbnail) {
    return (
      <img
        src={resizeContentfulImage(
          mediaItem.thumbnail.url,
          THUMBNAIL_MAX_WIDTH
        )}
        className={styles.thumbnail}
        alt={mediaItem.thumbnail.title}
      />
    );
  }
  const externalMediaUrl = mediaItem.url;
  const mediaItemClass = mediaItem.mediaItemClass;
  if (externalMediaUrl?.includes("vimeo")) {
    // Vimeo doesn't provide a thumbnail url without making an additional request to Vimeo
    // rendering an iframe without controls to avoid making the extra request
    return (
      <VimeoIFrameThumbnail
        videoUrl={externalMediaUrl}
        optanonClass={optanonClass}
        mediaItemClass={mediaItemClass}
      />
    );
  }
  // fallback icon
  return (
    <Icon
      source={BMI}
      style={{ width: 50, display: "block" }}
      color="primary"
    />
  );
};
