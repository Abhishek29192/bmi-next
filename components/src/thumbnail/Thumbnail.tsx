import { ButtonBase, ButtonBaseProps } from "@material-ui/core";
import classnames from "classnames";
import React from "react";
import { withClickable } from "../clickable/Clickable";
import Icon, { iconMap } from "../icon";
import Media, { AcceptedNode } from "../media/Media";
import { useStyles } from "./styles";

export enum StateEnum {
  ENABLED = "enabled",
  DISABLED = "disabled",
  SELECTED = "selected"
}

export enum SizeEnum {
  SMALL = "small",
  LARGE = "large"
}

export type Props = ButtonBaseProps & {
  media?: React.ReactElement<AcceptedNode>;
  altText?: string;
  state?: StateEnum;
  imageSource?: string;
  color?: string;
  size?: SizeEnum;
  isVideo: boolean;
  openYoutubeVideo?: (e: React.MouseEvent<SVGElement>) => void;
  visualiserParameters?: Record<string, string | number>;
};

const Thumbnail = ({
  media,
  altText,
  imageSource,
  color,
  state = StateEnum.ENABLED,
  size = SizeEnum.SMALL,
  className,
  isVideo = false,
  openYoutubeVideo,
  visualiserParameters,
  ...rest
}: Props) => {
  const classes = useStyles();
  const backgroundImage =
    (imageSource && isVideo) || (imageSource && !media)
      ? { backgroundImage: `url("${imageSource}")` }
      : {};

  const accessibilityText =
    (altText && isVideo) || (altText && !media) ? (
      <span className={classes.accessibilityText}>{altText}</span>
    ) : null;

  return (
    <ButtonBase
      disabled={state === StateEnum.DISABLED}
      disableRipple={state === StateEnum.SELECTED}
      className={classnames(
        className,
        classes.root,
        // eslint-disable-next-line security/detect-object-injection
        state !== StateEnum.ENABLED && classes[state],
        size === SizeEnum.LARGE && classes.large,
        state === StateEnum.SELECTED && classes.selected
      )}
      {...rest}
      style={{
        backgroundColor: color,
        ...backgroundImage
      }}
      data-testid={"default-thumbnail"}
    >
      {accessibilityText}
      {media && !isVideo && <Media>{media}</Media>}
      {isVideo && (
        <Icon
          source={iconMap.PlayArrow}
          className={classes.playIcon}
          onClick={(e) => openYoutubeVideo && openYoutubeVideo(e)}
          data-testid={"thumbnail-play-icon"}
        />
      )}
      {visualiserParameters && (
        <Icon source={iconMap.Cube} className={classes.cubeIcon} name="Cube" />
      )}
    </ButtonBase>
  );
};

export default withClickable(Thumbnail);
