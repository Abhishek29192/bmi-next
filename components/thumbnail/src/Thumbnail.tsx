import React from "react";
import classnames from "classnames";
import { withClickable } from "@bmi/clickable";
import { ButtonBase, ButtonBaseProps } from "@material-ui/core";
import Media, { AcceptedNode } from "@bmi/media";
import Icon, { iconMap } from "@bmi/icon";
import styles from "./Thumbnail.module.scss";

export enum StateEnum {
  ENABLED = "enabled",
  DISABLED = "disabled",
  SELECTED = "selected"
}
export enum SizeEnum {
  SMALL = "small",
  LARGE = "large"
}

export type Props = ButtonBaseProps & {
  media?: React.ReactElement<AcceptedNode>;
  altText?: string;
  state?: StateEnum;
  imageSource?: string;
  color?: string;
  size?: SizeEnum;
  isVideo: boolean;
};

const Thumbnail = ({
  media,
  altText,
  imageSource,
  color,
  state = StateEnum.ENABLED,
  size = SizeEnum.SMALL,
  className,
  isVideo = false,
  ...rest
}: Props) => {
  const classList = classnames(
    className,
    styles["Thumbnail"],
    state !== StateEnum.ENABLED && styles[`Thumbnail--${state}`],
    size === SizeEnum.LARGE && styles["Thumbnail--large"]
  );
  return (
    <ButtonBase
      disabled={state === StateEnum.DISABLED}
      disableRipple={state === StateEnum.SELECTED}
      className={classList}
      {...rest}
      style={{
        backgroundColor: color,
        backgroundImage: imageSource && `url("${imageSource}")`
      }}
    >
      {altText && (
        <span className={styles["accessibility-text"]}>{altText}</span>
      )}
      {media && <Media>{media}</Media>}
      <span className={styles["accessibility-text"]}>{altText}</span>
      {isVideo && (
        <Icon source={iconMap.PlayArrow} className={styles["play-icon"]} />
      )}
    </ButtonBase>
  );
};

export default withClickable(Thumbnail);
