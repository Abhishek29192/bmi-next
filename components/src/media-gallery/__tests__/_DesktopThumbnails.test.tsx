import { fireEvent, render } from "@testing-library/react";
import React from "react";
import Thumbnail from "../../thumbnail/Thumbnail";
import DesktopThumbnails from "../_DesktopThumbnails";
import mockImage from "./images/demo-tiles.jpg";

const onThumbnailClick = jest.fn();
const THUMBNAIL_WIDTH = 86;
export const getImages = (isMedia: boolean) => [
  {
    media: isMedia ? <img src={mockImage} alt="Lorem ipsum 1" /> : undefined,
    thumbnail: mockImage,
    altText: "test alt text 1",
    isVideo: false
  },
  {
    media: isMedia ? <img src={mockImage} alt="Lorem ipsum 2" /> : undefined,
    altText: "test alt text 2",
    thumbnail: mockImage,
    isVideo: false
  }
];
const renderedComponent = (isMedia = false) =>
  render(
    <DesktopThumbnails
      media={getImages(isMedia)}
      component={(props) => (
        <Thumbnail data-testid="default-thumbnail" {...props} />
      )}
      activeImageIndex={1}
      onThumbnailClick={onThumbnailClick}
    />
  );

beforeAll(() => {
  jest.clearAllMocks();
});

describe("_DesktopThumbnails component", () => {
  it("renders correctly without media prop", async () => {
    const images = getImages(false);
    const { container, getAllByTestId, findByText } = renderedComponent();

    const thumbnails = getAllByTestId("default-thumbnail");
    const scrollerButtons = container.querySelectorAll(".thumb-scroller");
    expect(container.firstChild).toMatchSnapshot();
    expect(scrollerButtons).toHaveLength(2);
    expect(thumbnails).toHaveLength(2);
    expect(thumbnails[1]).toHaveClass("Thumbnail--selected");
    expect(await findByText(images[0].altText as string)).toHaveClass(
      "accessibility-text"
    );

    fireEvent.click(thumbnails[1]);
    expect(onThumbnailClick).toHaveBeenCalledWith(expect.any(Object), 1);
  });
  it("renders correctly with media prop", async () => {
    const { container, getAllByTestId } = renderedComponent(true);

    const thumbnails = getAllByTestId("default-thumbnail");
    const scrollerButtons = container.querySelectorAll(".thumb-scroller");
    expect(container.firstChild).toMatchSnapshot();
    expect(scrollerButtons).toHaveLength(2);
    expect(thumbnails).toHaveLength(2);
    expect(thumbnails[1]).toHaveClass("Thumbnail--selected");
    expect(container.getElementsByTagName("img")).toHaveLength(2);

    fireEvent.click(thumbnails[1]);
    expect(onThumbnailClick).toHaveBeenCalledWith(expect.any(Object), 1);
  });

  describe("when click on the ThumbScrollerButton", () => {
    const parentOffsetWidth = 200;
    const images = getImages(false);
    const expectedMarginRight =
      -(THUMBNAIL_WIDTH * images.length - parentOffsetWidth) / 2;

    it("should scroll correctly", async () => {
      const { container } = renderedComponent();
      const scrollerButtonLeft = container.querySelector(
        ".thumb-scroller--left"
      ) as Element;
      const scrollerButtonRight = container.querySelector(
        ".thumb-scroller--right"
      ) as Element;
      const scroller = container.querySelector(".scroller") as Element;
      const parentElement = scroller?.parentElement as HTMLElement;

      Object.defineProperty(parentElement, "offsetWidth", {
        configurable: true,
        value: parentOffsetWidth
      });
      expect(container.querySelector(".scroller")).toHaveStyle(
        "margin-right: 0%"
      );

      fireEvent.click(scrollerButtonLeft);
      expect(scroller).toHaveStyle(`margin-right: ${expectedMarginRight}%`);

      fireEvent.transitionEnd(scroller);
      fireEvent.click(scrollerButtonRight);
      expect(scroller).toHaveStyle(`margin-right: ${expectedMarginRight * 2}%`);
    });

    it("should not scroll if isTransitioning", async () => {
      const { container } = renderedComponent();
      const scrollerButtonLeft = container.querySelector(
        ".thumb-scroller--left"
      ) as Element;
      const scrollerButtonRight = container.querySelector(
        ".thumb-scroller--right"
      ) as Element;
      const scroller = container.querySelector(".scroller") as Element;
      const parentElement = scroller?.parentElement as HTMLElement;

      Object.defineProperty(parentElement, "offsetWidth", {
        configurable: true,
        value: parentOffsetWidth
      });

      fireEvent.click(scrollerButtonLeft);
      expect(scroller).toHaveStyle(`margin-right: ${expectedMarginRight}%`);

      fireEvent.click(scrollerButtonRight);
      expect(scroller).toHaveStyle(`margin-right: ${expectedMarginRight}%`);
    });

    it("should scroll correctly when no parentElement offsetWidth", () => {
      const { container } = renderedComponent();
      const scrollerButtonLeft = container.querySelector(
        ".thumb-scroller--left"
      ) as Element;
      const scrollerButtonRight = container.querySelector(
        ".thumb-scroller--right"
      ) as Element;
      const scroller = container.querySelector(".scroller") as Element;

      expect(container.querySelector(".scroller")).toHaveStyle(
        "margin-right: 0%"
      );

      fireEvent.click(scrollerButtonLeft);
      expect(scroller).toHaveStyle(`margin-right: -100%`);

      fireEvent.transitionEnd(scroller);
      fireEvent.click(scrollerButtonRight);
      expect(scroller).toHaveStyle(`margin-right: 0%`);
    });
  });

  it("shoule run calculateLeftRightVisibility correctly when thumbnailsElement offsetWidth is not 0 or undefined", async () => {
    const { container } = renderedComponent();
    const scroller = container.querySelector(".scroller") as Element;
    const scrollerButtonLeft = container.querySelector(
      ".thumb-scroller--left"
    ) as Element;

    Object.defineProperty(scroller, "offsetWidth", {
      configurable: true,
      value: 50
    });

    fireEvent.click(scrollerButtonLeft);
    expect(
      container.querySelector(".thumb-scroller--left.thumb-scroller--hidden")
    ).toBeFalsy();
    expect(
      container.querySelector(".thumb-scroller--right.thumb-scroller--hidden")
    ).toBeTruthy();

    fireEvent.transitionEnd(scroller);

    expect(
      container.querySelector(".thumb-scroller--left.thumb-scroller--hidden")
    ).toBeFalsy();
    expect(
      container.querySelector(".thumb-scroller--right.thumb-scroller--hidden")
    ).toBeFalsy();
  });

  it("should run calculateLeftRightVisibility correctly when no parentElement", async () => {
    const { container } = renderedComponent();
    const scrollerButtonLeft = container.querySelector(
      ".thumb-scroller--left"
    ) as Element;
    const scroller = container.querySelector(".scroller") as Element;

    Object.defineProperty(scroller, "offsetWidth", {
      configurable: true,
      value: 50
    });
    fireEvent.click(scrollerButtonLeft);
    Object.defineProperty(scroller, "parentElement", {
      configurable: true,
      value: null
    });
    Object.defineProperty(scroller, "offsetWidth", {
      configurable: true,
      value: 0
    });

    expect(
      container.querySelector(".thumb-scroller--left.thumb-scroller--hidden")
    ).toBeFalsy();
    expect(
      container.querySelector(".thumb-scroller--right.thumb-scroller--hidden")
    ).toBeTruthy();

    fireEvent.transitionEnd(scroller);

    expect(
      container.querySelector(".thumb-scroller--left.thumb-scroller--hidden")
    ).toBeFalsy();
    expect(
      container.querySelector(".thumb-scroller--right.thumb-scroller--hidden")
    ).toBeTruthy();
  });
});
