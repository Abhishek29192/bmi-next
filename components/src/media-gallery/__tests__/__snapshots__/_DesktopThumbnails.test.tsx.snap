// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`_DesktopThumbnails component renders correctly with media prop 1`] = `
<div
  class="Thumbnails-root-22"
  data-testid="thumbnails-root"
>
  <button
    aria-label="Scroll left"
    class="MuiButtonBase-root Thumbnails-thumbScroller-23 Thumbnails-thumbScrollerLeft-24 ThumbScrollerButton-root-32 ThumbScrollerButton-left-33"
    data-testid="thumbnail-scroller-left"
    tabindex="-1"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root ThumbScrollerButton-icon-35"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
      />
    </svg>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
  <button
    aria-label="Scroll right"
    class="MuiButtonBase-root Thumbnails-thumbScroller-23 Thumbnails-thumbScrollerRight-25 Thumbnails-thumbScrollerHidden-26 ThumbScrollerButton-root-32 ThumbScrollerButton-right-34"
    data-testid="thumbnail-scroller-right"
    tabindex="-1"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root ThumbScrollerButton-icon-35"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
      />
    </svg>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
  <div
    class="Thumbnails-scroller-27"
    data-testid="thumbnail-scroller"
    style="margin-right: 0%;"
  >
    <button
      class="MuiButtonBase-root Thumbnails-thumbnail-28 Thumbnail-root-36 Thumbnail-large-39"
      data-testid="default-thumbnail"
      tabindex="0"
      type="button"
    >
      <img
        alt="Lorem ipsum 1"
        class="Media-root-43"
        src="path-to-image"
      />
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <button
      class="MuiButtonBase-root Thumbnails-thumbnail-28 Thumbnail-root-36 Thumbnail-selected-38 Thumbnail-large-39 Thumbnail-selected-38"
      data-testid="default-thumbnail"
      tabindex="0"
      type="button"
    >
      <img
        alt="Lorem ipsum 2"
        class="Media-root-43"
        src="path-to-image"
      />
    </button>
  </div>
</div>
`;

exports[`_DesktopThumbnails component renders correctly without media prop 1`] = `
<div
  class="Thumbnails-root-1"
  data-testid="thumbnails-root"
>
  <button
    aria-label="Scroll left"
    class="MuiButtonBase-root Thumbnails-thumbScroller-2 Thumbnails-thumbScrollerLeft-3 ThumbScrollerButton-root-11 ThumbScrollerButton-left-12"
    data-testid="thumbnail-scroller-left"
    tabindex="-1"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root ThumbScrollerButton-icon-14"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
      />
    </svg>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
  <button
    aria-label="Scroll right"
    class="MuiButtonBase-root Thumbnails-thumbScroller-2 Thumbnails-thumbScrollerRight-4 Thumbnails-thumbScrollerHidden-5 ThumbScrollerButton-root-11 ThumbScrollerButton-right-13"
    data-testid="thumbnail-scroller-right"
    tabindex="-1"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root ThumbScrollerButton-icon-14"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
      />
    </svg>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
  <div
    class="Thumbnails-scroller-6"
    data-testid="thumbnail-scroller"
    style="margin-right: 0%;"
  >
    <button
      class="MuiButtonBase-root Thumbnails-thumbnail-7 Thumbnail-root-15 Thumbnail-large-18"
      data-testid="default-thumbnail"
      style="background-image: url(path-to-image);"
      tabindex="0"
      type="button"
    >
      <span
        class="Thumbnail-accessibilityText-19"
      >
        test alt text 1
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <button
      class="MuiButtonBase-root Thumbnails-thumbnail-7 Thumbnail-root-15 Thumbnail-selected-17 Thumbnail-large-18 Thumbnail-selected-17"
      data-testid="default-thumbnail"
      style="background-image: url(path-to-image);"
      tabindex="0"
      type="button"
    >
      <span
        class="Thumbnail-accessibilityText-19"
      >
        test alt text 2
      </span>
    </button>
  </div>
</div>
`;
